### ISSUE TITLE
onEdgeUpdate crashes when using CustomEdge and loose connectionMode

### ISSUE THREAD

**Issue by obaid** [wbkd/react-flow#982]
So I am not sure if it is an issue with my setup but if I set the connectionMode to `loose` and try to drag  the edge from one node to another I get the following issue:

![Screen Shot 2021-03-09 at 9 13 21 AM](https://user-images.githubusercontent.com/12336/110483511-c2c51b00-80b7-11eb-90c6-a1c833bfc1d7.png)

It is occurring because the `connectionHandleType` for `target` is a null array. I think it should be an empty array instead.

![Screen Shot 2021-03-09 at 9 12 56 AM](https://user-images.githubusercontent.com/12336/110483490-bf319400-80b7-11eb-8950-6799bab5393b.png)

**Comment by obaid** [comment]
Yes, and I found the issue is fixed by using `useUpdateNodeInternals` .. so going to close this one :)

**Comment by obaid** [comment]
Actually I spoke too early. :(

The issue does exists even after called `useUpdateNodeInternals` the target is still undefined.

**Comment by moklick** [comment]
When all your handles have `type="source"` it should look into `handleBounds.source` array. Can you post your CustomNode implementation here?

**Comment by obaid** [comment]
In my custom node:

```
const CustomNode = (props) => {
  return (
    <>
      <div style={{ width: "150px", height: "30px" }}>Custom Node</div>
      <Handle
        type="source"
        position="right"
        id="1"
        style={{ background: "#555" }}
      />
      <Handle
        type="source"
        position="bottom"
        id="2"
        style={{ background: "#555" }}
      />
      <Handle
        type="source"
        position="top"
        id="3"
        style={{ background: "#555" }}
      />
      <Handle
        type="source"
        position="left"
        id="4"
        style={{ background: "#555" }}
      />
    </>
  );
};
```

**Comment by obaid** [comment]
Shouldn't this return an empty array instead of `null`?

https://github.com/wbkd/react-flow/blob/42019a6a6659b201beefb873d950db2983493fdf/src/components/Nodes/utils.ts#L22

**Comment by moklick** [comment]
Why is `connectionHandleType = "target"` in your case?

**Comment by obaid** [comment]
I believe this is where it is being set to "target':

 https://github.com/wbkd/react-flow/blob/e9f97566fbf82e262e0f60115478417377970c02/src/components/Handle/handler.ts#L106

**Comment by obaid** [comment]
So you can see a reproducible example here: https://codesandbox.io/s/confident-chebyshev-re3w3?file=/src/CustomNode.js

The CustomNode is exactly the same as the example (I only added node id to the div). If you create a connection between any two block and then try to move that connector it will crash. See gif below:

![bug](https://user-images.githubusercontent.com/12336/110511448-3d9b2f80-80d2-11eb-8890-f6b3ea49b0ad.gif)

**Comment by moklick** [comment]
I should have read the title of this issue more carefully ;) It's about the `onEdgeUpdate`.. I will check that. It seems like a bug.

**Comment by clementloridan** [comment]
I think the `connectionMode` feature was not implement in all code. No worries we will fix. But a little documentation about the connectionMode loose is needed.

- Here, `connectionHandleType` can be `source` or `target`.

https://github.com/wbkd/react-flow/blob/e9f97566fbf82e262e0f60115478417377970c02/src/container/EdgeRenderer/index.tsx#L176

- And here, the logic follow than it will check the handleBounds linked to the type of the handle: `target` or `source`. 
https://github.com/wbkd/react-flow/blob/e9f97566fbf82e262e0f60115478417377970c02/src/components/ConnectionLine/index.tsx#L57

I don't have enough information about the connectionMode loose to give a smart solution.

Because it's can be from here:
https://github.com/wbkd/react-flow/blob/e9f97566fbf82e262e0f60115478417377970c02/src/components/Nodes/utils.ts#L9

But I think it's better to rethinking about the connectionMode loose to remove all the logic of source and target because this is goal of this mode.

**Comment by moklick** [comment]
The fix is released in v9.2.0.

I am not super happy with the fix but it works. I will rethink the current handle/edge behaviour and hopefully come up with a more robust solution.
